src/core/p5.Renderer.js

```diff
@@ -289,23 +283,79 @@ p5.Renderer.prototype.text = function(str, x, y, maxWidth, maxHeight) {
           this._textBaseline = constants.TOP;
           break;
       }
-
       // remember the max-allowed y-position for any line (fix to #928)
       finalMaxHeight = y + maxHeight - p.textAscent();
     }
 
-    for (ii = 0; ii < cars.length; ii++) {
-      line = '';
-      words = cars[ii].split(' ');
-      for (n = 0; n < words.length; n++) {
-        testLine = `${line + words[n]} `;
-        testWidth = this.textWidth(testLine);
-        if (testWidth > maxWidth && line.length > 0) {
-          this._renderText(p, line, x, y, finalMaxHeight);
-          line = `${words[n]} `;
-          y += p.textLeading();
-        } else {
-          line = testLine;
+    // Render lines of text according to settings of textWrap and textHyphens
+    // Splits lines at spaces, for loop adds one word + space at a time and tests length with next word added. If line + next word is too long, print currents line
+    if (textWrapStyle === constants.LINE) {
+      for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
+        line = '';
+        words = lines[lineIndex].split(' ');
+        for (let wordIndex = 0; wordIndex < words.length; wordIndex++) {
+          testLine = `${line + words[wordIndex]} `;
+          testWidth = this.textWidth(testLine);
+          if (testWidth > maxWidth && line.length > 0) {
+            this._renderText(p, line, x, y, finalMaxHeight);
+            line = `${words[wordIndex]} `;
+            y += p.textLeading();
+          } else {
+            line = testLine;
+          }
+        }
+        this._renderText(p, line, x, y, finalMaxHeight);
+        y += p.textLeading();
+
+        if (baselineHacked) {
+          this._textBaseline = constants.BASELINE;
+        }
+      }
+    } else {
+      // Splits lines at characters, for loop adds one char at a time and tests length with next char added. If line + next char is too long, chars are trimmed according to additional hyphenation rules
```

Comment:
@dhowe thank you for taking the time to review this!

I'm a little confused that you referenced the commit message with the disappearing last line problem - I think you may have checked out from the wrong branch. I have now deleted the WIP branch `text-wrap-and-hyphens` to avoid confusion and would ask that you make sure you are checking out the current branch `add-text-wrap-and-hyphens` where I believe I already fixed this issue :) 

Otherwise it could be the case that you are seeing the normal text behavior - Text that does not fit completely within the rectangle specified will not be drawn to the screen. If you don't include the maxHeight parameter (h) I believe this line would not be cut off.

Reply:
Ah sorry, my bad - not sure how that happened... In any case, I have resolved this part of the review, but I think the other issues are still relevant. How does my suggestion re: wrapping options sound to you?

Reply:
Yes! We are definitely in agreement about textWrap. Let's discuss more downthread, I responded to it as a separate comment.